In devcontainer, tool installs fail

See below error. An analogous error happens for Rhino.

In theory, this may be because we are attempting to write to a Docker volume here. The same code seems to work inside GitHub Actions.

As a workaround, using the `mv` command after the error to move the file where it was trying to move it succeeds, for unknown reasons.

```
root@e54ca0f39833:/slime# ./fifty test.jsh rhino/shell/module.fifty.ts 
java.lang.RuntimeException: java.io.IOException: Failed to move: /tmp/slime3813334804388148718.tmp/node-v16.13.1-linux-x64 to /slime/local/jsh/lib/node
        at jdk.nashorn.internal.runtime.ScriptRuntime.apply(ScriptRuntime.java:397)
        at jdk.nashorn.internal.scripts.Script$Recompilation$33$4303AAAA$nashorn.L:19#nashorn#compile(slime://loader/jrunscript/nashorn.js:109)
        at jdk.nashorn.internal.scripts.Script$Recompilation$31$5603AAAA$nashorn.L:19#nashorn#script(slime://loader/jrunscript/nashorn.js:139)
        at jdk.nashorn.internal.scripts.Script$Recompilation$30$10914AAAA$nashorn.L:19#$javahost#script(slime://loader/jrunscript/nashorn.js:269)
        at jdk.nashorn.internal.scripts.Script$Recompilation$48$3642AAA$expression.L:18#slime#execute(slime://loader/jrunscript/expression.js:112)
        at jdk.nashorn.internal.scripts.Script$Recompilation$124$2631AA$scripts.L:19#run(slime://loader/scripts.js:131)
        at jdk.nashorn.internal.scripts.Script$Recompilation$120$7866AAA$expression.L:13#topMethod#L:331(slime://loader/expression.js:332)
        at jdk.nashorn.internal.scripts.Script$Recompilation$855$jsh.L:16#jsh#loader#run(/slime/jsh/loader/jsh.js:242)
        at jdk.nashorn.internal.scripts.Script$Recompilation$854$16139AAA$plugin_jsh.L:18#load-2#L:514#L:515#L:516(/slime/jsh/script/plugin.jsh.js:533)
        at jdk.nashorn.internal.scripts.Script$Recompilation$151$485AAAA$jsh.L:16(/slime/jsh/loader/jsh.js:416)
        at jdk.nashorn.internal.scripts.Script$150$jsh.:program(/slime/jsh/loader/jsh.js:8)
        at jdk.nashorn.internal.runtime.ScriptFunctionData.invoke(ScriptFunctionData.java:637)
        at jdk.nashorn.internal.runtime.ScriptFunction.invoke(ScriptFunction.java:494)
        at jdk.nashorn.internal.runtime.ScriptRuntime.apply(ScriptRuntime.java:393)
        at jdk.nashorn.api.scripting.NashornScriptEngine.evalImpl(NashornScriptEngine.java:449)
        at jdk.nashorn.api.scripting.NashornScriptEngine.evalImpl(NashornScriptEngine.java:406)
        at jdk.nashorn.api.scripting.NashornScriptEngine.evalImpl(NashornScriptEngine.java:402)
        at jdk.nashorn.api.scripting.NashornScriptEngine.eval(NashornScriptEngine.java:155)
        at inonit.script.engine.Host$JavaxScriptHost.eval(Host.java:95)
        at inonit.script.engine.Host.run(Host.java:118)
        at inonit.script.engine.Host.run(Host.java:127)
        at inonit.script.jsh.Shell$Execution$1.run(Shell.java:727)
        at inonit.script.jsh.Nashorn$ExecutionImpl$ErrorHandlingImpl.run(Nashorn.java:112)
        at inonit.script.jsh.Shell$Execution.execute(Shell.java:724)
        at inonit.script.jsh.Nashorn.main(Nashorn.java:40)
        at inonit.script.jsh.Shell$Engine.shell(Shell.java:646)
        at inonit.script.jsh.Main.cli(Main.java:536)
        at inonit.script.jsh.Nashorn.main(Nashorn.java:136)
Caused by: java.io.IOException: Failed to move: /tmp/slime3813334804388148718.tmp/node-v16.13.1-linux-x64 to /slime/local/jsh/lib/node
        at inonit.script.runtime.io.Filesystem$NativeFilesystem$NodeImpl.move(Filesystem.java:302)
        at jdk.nashorn.internal.scripts.Script$Recompilation$951$8171AA$java.L:17#FilesystemProvider#move(/slime/rhino/file/java.js:287)
        at jdk.nashorn.internal.scripts.Script$Recompilation$948$9547AA$file.L:16#Pathname#Node#move(/slime/rhino/file/file.js:342)
        at jdk.nashorn.internal.scripts.Script$Recompilation$884$4840AA$module.L:15#installLocalArchive(/slime/jrunscript/tools/install/module.js:175)
        at jdk.nashorn.internal.scripts.Script$Recompilation$881$6871AA$module.L:15#install(/slime/jrunscript/tools/install/module.js:200)
        at jdk.nashorn.internal.scripts.Script$Recompilation$880$7187AA$module.L:15#oldInstall(/slime/jrunscript/tools/install/module.js:204)
        at jdk.nashorn.internal.scripts.Script$Recompilation$863$4221AA$events.L:14#listening#L:161(slime://loader/events.js:174)
        at jdk.nashorn.internal.scripts.Script$Recompilation$878$8809AA$module.L:15#install-1(/slime/jrunscript/tools/install/module.js:258)
        at jdk.nashorn.internal.scripts.Script$Recompilation$877$7903AA$module.L:14#install(/slime/rhino/tools/node/module.js:275)
        at jdk.nashorn.internal.scripts.Script$Recompilation$863$4221AA$events.L:14#listening#L:161(slime://loader/events.js:174)
        at jdk.nashorn.internal.scripts.Script$Recompilation$874$21701A$plugin_jsh.L:16#load-1#integratedNode#install(/slime/jsh/tools/install/plugin.jsh.js:712)
        at jdk.nashorn.internal.scripts.Script$Recompilation$873$21374$plugin_jsh.L:16#load-1#integratedNode#require#install(/slime/jsh/tools/install/plugin.jsh.js:693)
        at jdk.nashorn.internal.scripts.Script$Recompilation$867$17567AA$jsh.L:17#require(/slime/rhino/shell/jsh.js:542)
        at jdk.nashorn.internal.scripts.Script$Recompilation$863$4221AA$events.L:14#listening#L:161(slime://loader/events.js:174)
        at jdk.nashorn.internal.scripts.Script$Recompilation$862$21218$plugin_jsh.L:16#load-1#integratedNode#require(/slime/jsh/tools/install/plugin.jsh.js:688)
        at jdk.nashorn.internal.scripts.Script$Recompilation$861$16497A$plugin_jsh.L:16#load#typescript#require(/slime/tools/wf/plugin.jsh.js:520)
        at jdk.nashorn.internal.scripts.Script$Recompilation$860$347AA$test_jsh.L:13(/slime/tools/fifty/test.jsh.js:14)
        at jdk.nashorn.internal.scripts.Script$859$test_jsh.:program(/slime/tools/fifty/test.jsh.js:8)
        at jdk.nashorn.internal.runtime.ScriptFunctionData.invoke(ScriptFunctionData.java:637)
        at jdk.nashorn.internal.runtime.ScriptFunction.invoke(ScriptFunction.java:494)
        at jdk.nashorn.internal.runtime.ScriptRuntime.apply(ScriptRuntime.java:393)
        ... 27 more
```

--- a/.devcontainer/devcontainer.json
+++ b/.devcontainer/devcontainer.json
@@ -15,7 +15,7 @@
 		"dbaeumer.vscode-eslint"
 	],
 	"settings": {
-		"java.jdt.ls.java.home": "/slime/local/jdk/11"
+		"java.jdt.ls.java.home": "/slime/local/jdk/17"
 	},
 	"service": "test",
 	"workspaceFolder": "/slime",

--- a/.devcontainer/post-create-command.bash
+++ b/.devcontainer/post-create-command.bash
@@ -5,4 +5,4 @@
 #	END LICENSE
 
 ./wf initialize
-./jsh.bash --add-jdk-11
+./jsh.bash --add-jdk-17

--- a/contributor/docker-compose-run
+++ b/contributor/docker-compose-run
@@ -6,5 +6,5 @@
 #	END LICENSE
 
 SLIME="$(dirname $0)/.."
-docker compose -f $SLIME/contributor/docker-compose.yaml build slime \
-	&& docker compose -f $SLIME/contributor/docker-compose.yaml run -w /slime slime /bin/bash "$@"
+docker compose -f $SLIME/contributor/docker-compose.yaml build test \
+	&& docker compose -f $SLIME/contributor/docker-compose.yaml run -w /slime test /bin/bash "$@"

--- a/jrunscript/tools/install/module.js
+++ b/jrunscript/tools/install/module.js
@@ -172,10 +172,20 @@
 			var unzippedTo = untardir.getSubdirectory(unzippedDestination);
 			if (!unzippedTo) throw new TypeError("Expected directory " + unzippedDestination + " not found in " + untardir);
 			events.fire("console", "Directory is: " + unzippedTo);
-			unzippedTo.move(p.to, {
-				overwrite: p.replace,
-				recursive: true
-			});
+			//	TODO	right now, we will use the mv command preferentially because it works in some situations the Java
+			//			java.io.File renameTo implementation does not (notable on our Docker setup, moving from a temporary
+			//			directory to the installation directory, as we are doing here).
+			if ($context.api.shell.PATH.getCommand("mv")) {
+				$context.api.shell.run({
+					command: "mv",
+					arguments: [unzippedTo.pathname.toString(), p.to.toString()]
+				});
+			} else {
+				unzippedTo.move(p.to, {
+					overwrite: p.replace,
+					recursive: true
+				});
+			}
 			return p.to.directory;
 		};
 

--- a/jrunscript/tools/install/module.js
+++ b/jrunscript/tools/install/module.js
@@ -176,6 +176,11 @@
 			//			java.io.File renameTo implementation does not (notable on our Docker setup, moving from a temporary
 			//			directory to the installation directory, as we are doing here).
 			if ($context.api.shell.PATH.getCommand("mv")) {
+				p.to.parent.createDirectory({
+					exists: function(dir) {
+						return false;
+					}
+				});
 				$context.api.shell.run({
 					command: "mv",
 					arguments: [unzippedTo.pathname.toString(), p.to.toString()]

--- a/jsh.bash
+++ b/jsh.bash
@@ -164,9 +164,12 @@ install_jdk_8_corretto() {
 	install_jdk_corretto "8.332.08.1" "8" $1
 }
 
-#	TODO	a lot of copy-paste with install_jdk_8_corretto
 install_jdk_11_corretto() {
 	install_jdk_corretto "11.0.15.9.1" "11" $1
+}
+
+install_jdk_17_corretto() {
+	install_jdk_corretto "17.0.3.6.1" "17" $1
 }
 
 install_jdk_8() {
@@ -175,6 +178,10 @@ install_jdk_8() {
 
 install_jdk_11() {
 	install_jdk_11_corretto "$@"
+}
+
+install_jdk_17() {
+	install_jdk_17_corretto "$@"
 }
 
 install_jdk() {
@@ -198,6 +205,11 @@ fi
 
 if [ "$1" == "--add-jdk-11" ]; then
 	install_jdk_11 ${JSH_LOCAL_JDKS}/11
+	exit $?
+fi
+
+if [ "$1" == "--add-jdk-17" ]; then
+	install_jdk_17 ${JSH_LOCAL_JDKS}/17
 	exit $?
 fi
 

--- a/rhino/file/java/inonit/script/runtime/io/Filesystem.java
+++ b/rhino/file/java/inonit/script/runtime/io/Filesystem.java
@@ -7,6 +7,7 @@
 package inonit.script.runtime.io;
 
 import java.io.*;
+import java.nio.file.*;
 import java.util.logging.*;
 
 import inonit.system.*;
@@ -296,10 +297,28 @@ public abstract class Filesystem {
 				if (!success) throw new IOException("Failed to delete: " + this.file);
 			}
 
+			private void copy(File from, File to) throws IOException {
+				if (from.isDirectory()) {
+					File[] files = from.listFiles();
+					for (File file : files) {
+						copy(file, new File(to, file.getName()));
+					}
+				} else {
+					if (!to.getParentFile().exists()) {
+						to.getParentFile().mkdirs();
+					}
+					Files.copy(from.toPath(), to.toPath(), StandardCopyOption.COPY_ATTRIBUTES);
+				}
+			}
+
 			public void move(Node to) throws IOException {
 				NodeImpl toNode = (NodeImpl)to;
 				boolean success = file.renameTo(toNode.file);
-				if (!success) throw new IOException("Failed to move: " + this.file + " to " + toNode.file);
+				if (!success) {
+					//	TODO	this does not work for symlinks; it appears to copy them "by value"
+					copy(file, toNode.file);
+					file.delete();
+				}
 			}
 
 			public void mkdir() throws IOException {

--- a/rhino/file/java/inonit/script/runtime/io/Filesystem.java
+++ b/rhino/file/java/inonit/script/runtime/io/Filesystem.java
@@ -304,9 +304,6 @@ public abstract class Filesystem {
 						copy(file, new File(to, file.getName()));
 					}
 				} else {
-					if (!to.getParentFile().exists()) {
-						to.getParentFile().mkdirs();
-					}
 					Files.copy(from.toPath(), to.toPath(), StandardCopyOption.COPY_ATTRIBUTES);
 				}
 			}

--- a/tools/fifty/test.jsh.js
+++ b/tools/fifty/test.jsh.js
@@ -11,7 +11,29 @@
 	 * @param { slime.jsh.Global } jsh
 	 */
 	function($api,jsh) {
-		jsh.wf.typescript.require();
+		var isTypescriptInstalled = function() {
+			var installation = jsh.shell.tools.node.installation;
+			var nodeExists = $api.Function.world.ask(
+				jsh.shell.tools.node.world.Installation.exists(installation)
+			)();
+			if (!nodeExists) return false;
+			var typescript = jsh.shell.tools.node.world.Installation.modules.installed("typescript");
+			var tsInstalled = $api.Function.world.now.question(
+				typescript,
+				installation
+			);
+			//	TODO	this simply ensures that *some* version of TypeScript is installed.
+			return tsInstalled.present;
+		}
+
+		//	We need to use this method, which forks a new shell, because we need TypeScript in this running shell in order to load
+		//	the Fifty tests. May want to provide this as an API somewhere (currently there is one in jsh.wf, but it functions
+		//	slightly differently; a `jsh` script requiring TypeScript is very foreseeable, though). Could generalize to require a
+		//	specific TypeScript version, etc.
+		jsh.shell.jsh.require({
+			satisfied: isTypescriptInstalled,
+			install: function() { jsh.wf.typescript.require(); }
+		});
 
 		/** @type { slime.jsh.script.cli.Processor<any, { definition: slime.jrunscript.file.File, list: boolean, part: string, view: string }> } */
 		var processor = $api.Function.pipe(

--- a/tools/wf/plugin.jsh.js
+++ b/tools/wf/plugin.jsh.js
@@ -16,7 +16,7 @@
 	function($api,$slime,jsh,$loader,plugin) {
 		plugin({
 			isReady: function() {
-				return Boolean(jsh.file && jsh.shell && jsh.ui && jsh.tools && jsh.tools.git);
+				return Boolean(jsh.file && jsh.shell && jsh.shell.tools && jsh.ui && jsh.tools && jsh.tools.git);
 			},
 			load: function() {
 				jsh.wf = {

